### DISCUSSION TITLE
TestStore Changing Environment Variable

### DISCUSSION THREAD

**Discussion by natemann** [pointfreeco/swift-composable-architecture#490] [Q&A] [answered]
I have an effect that attempts to download data every 60 seconds
```swift

public struct Environment {
        var backgroundQueue: AnySchedulerOf<DispatchQueue>
        var download: () -> AnyPublisher<Data, APIError>
        var logError: (Error) -> Effect<Never, Never>
        var now: () -> Date
        var runLoop: AnySchedulerOf<RunLoop>
        var userDefaults: UserDefaults
}
````

```swift
case .start:
Effect.timer(id: CancelTimer(), every: .seconds(60), on: environment.runLoop)
                .prepend(.init(environment.now()))
                .assertNoFailure()
                .flatMap(maxPublishers: .max(1), { _ -> Effect<Result<Data, APIError>, Never> in
                    environment.download()
                        .handleEvents(receiveOutput: { _ in print("Downloading data") })
                        .catchToEffect()
                })
                .map(Action.save)
                .eraseToEffect()
                .cancellable(id: CancelDownload())
```

because the download can take some time, I limit the download to 1 in progress with the `flatMap(maxPublishers`. Everything works fine in production, but I am trying to test that when a download error is received, and then 60 seconds later, data is received, everything continues to work. Here is my test

```swift
func test_download_failure_then_success() {
        let runLoop = RunLoop.test
        var errors: [Error] = []
        let store = TestStore(
            initialState: Downloader.State(),
            reducer: Downloader.reducer,
            environment: Downloader.Environment(
                backgroundQueue: DispatchQueue.test.eraseToAnyScheduler(),
                download: { Fail(error: APIError.invalidBody).eraseToAnyPublisher() },
                logError: { error in .fireAndForget { errors.append(error) } },
                now: Date.init,
                runLoop: runLoop.eraseToAnyScheduler(),
                userDefaults: UserDefaults.mock()))

        store.assert(
            .send(.start),
            .receive(.save(.failure(.invalidBody))),
            .environment {
                $0.download = { Just(Data()).mapError(APIError.unknown).eraseToAnyPublisher() }
            },

            .do { runLoop.advance(by: .seconds(60)) },
            .receive(.save(.success(.init()))),
            .send(.cancel)
        )
```

The environment.download success is never called. I imagine this has something to do with the maxPublishers, but I cannot figure it out. Any help would be appreciated.

**Comment by mbrandonw** [comment, marked as answer]
I think I see what is going on here, and it's tricky.

The change you make to the environment to replace the `$0.download` endpoint doesn't actually update the environment in the `flatMap`. The `flatMap` will capture the environment at the time of sending the `.start` action, and will never change after that. In fact, that kinda makes using `environment` inside `.flatMaps` a little dangerous, in particular when flat mapping from a long living effect like a timer. We should add some documentation/case studies for this behavior.

One thing you should be able to do to work around is use a `PassthroughSubject` for `environment.download` in your test and then send it values to emulate the `download` effect firing, rather than replacing the environment.

Wanna give that a shot?

There's another work around for this but it won't work for you because you specifically need the `maxPublishers` behavior. But the idea is to create a new action that can be sent from the effect in the `.flatMap` and then do your work inside that action. That extra little ping-pong with actions gives you back the ability to update the environment while the timer is running.

**Comment by natemann** [comment]
@mbrandonw thanks for the help. Using a `PassThroughSubject` worked after I changed download to `Effect<Result<Data, APIError>, Never>`
